example_code/shmem_ctx_pipelined_reduce.c

```diff
@@ -0,0 +1,56 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <shmem.h>
+
+#define LEN  8192 /* Full buffer length */
+#define PLEN 512  /* Length of each pipeline stage */
+
+int in_buf[LEN], out_buf[LEN];
+
+int main(void) {
+    int i, j, *pbuf[2];
+    shmem_ctx_t ctx[2];
+
+    shmem_init();
+    int me = shmem_my_pe();
+    int npes = shmem_n_pes();
+
+    pbuf[0] = shmem_malloc(PLEN * npes * sizeof(int));
+    pbuf[1] = shmem_malloc(PLEN * npes * sizeof(int));
+
+    int ret_0 = shmem_ctx_create(0, &ctx[0]);
+    int ret_1 = shmem_ctx_create(0, &ctx[1]);
```

Comment:
> Observation: Private contexts would allow sync be replaced with barrier without significant performance penalty.

FWIU sync can be implemented on a separate unique resource which is completely independent of the progress on resources mapped to any SHARED/PRIVATE/DEFAULT context. If that is the case, then there will be significant performance difference.